### DISCUSSION TITLE
Wasm support and custom client

### DISCUSSION THREAD

**Discussion by arn-the-long-beard** [Byron/google-apis-rs#279] [Ideas]
Hello guys ! 
First thank you so much for your awesome package of generated Apis with Rust  :heart: . That is gonna be so useful :grin: .

But I was wondering about some ideas regarding the support for WASM. I actually do need to use a custom web client instead of hyper for requesting data. I do not wish to do it on backend because I am concerned for having way too much load on it. ( or is there some strategy for it  ?)

I was wondering : 

- 1. Are you interested/planning to have support for WASM ?
- 2. Is it too much work to use `feature gate` for custom client for example to enable other clients to be used with the Api ( I am thinking for youtube Api in my case)
- 3. Do you think I should actually do my own implementation on the front end for WASM, by getting inspiration from the [official javascript client](https://github.com/google/google-api-javascript-client) and your generated code?

Any advice would be welcome :wink: , I am not familiar with Apis right now, So that is a new world I am jumping in :laughing:  .

**Comment by Byron** [comment]
Thanks for your interest in the project! As it's in [maintenance mode](https://github.com/Byron/google-apis-rs#maintenance-mode), I will not implement new features but PRs are always welcome, too. This probably answers `i`.

Regarding `ii`, as long as the client uses [http-types](https://docs.rs/http/0.2.4/http/) it should be relatively straightforward to support non-hyper clients. The devil might be in the details though.

For `iii`, that's probably for you to decide once you had a look at everything. Definitely include [the next generation of Google API crates](https://github.com/google-apis-rs) in your exploration - they might already do what you need or they might be easier to contribute to.

I hope that helps.

**Comment by arn-the-long-beard** [comment]
Hello ! 
Thank you for such quick answer !

Yes it uses http-types :wink: 

And thank you so much for the link to next projects :grin:  , I will continue my exploration !

**Comment by arn-the-long-beard** [comment]
But I have one question by the way. Why do we need to generate the Apis  instead of having ready to use libraries based on google documentation for implementation  https://developers.google.com/youtube/v3/guides/implementation ?

I am very fresh to this context and I am wondering :thinking:

**Comment by Byron** [comment]
I don't think I understand the question. What do you mean with 'ready to use libraries'?

Assuming you mean that libraries for other programming languages exist, then this project was created to do the same for Rust.

**Reply by arn-the-long-beard** [reply]
Hello,

Thank you for your message :smile: .

I did formulate my question the wrong way :facepalm: 

I have been contributing to a project for driver for a database Api and it was straight forward. We just needed to read the documentation and implement it.

But here for the google Apis, it seems we need to generate the Apis instead of coding it , is it right ? I would like to contribute, especially to the next implementations. But I have a hard time to understand the flow and were could I update the code to make it WASM compatible.

I thought I sent my message yesterday, Too busy days . tried today to just use directly the Apis from the explorer provided online by Google for youtube, and yes I understand that actually generating the data models with a tool is very helpful then :smile: .

**Comment by Byron** [comment]
> But here for the google Apis, it seems we need to generate the Apis instead of coding it , is it right ?

It's right. I recommend trying to ask for the feature over at [the next generation](https://github.com/google-apis-rs) or try to contribute it there. There is various generators even with [this one](https://github.com/google-apis-rs/google-cloud-rs) maybe the most active.

**Reply by arn-the-long-beard** [reply]
Thank you, I will do this . Thank you for the discussion !

**Comment by allsey87** [comment]
@arn-the-long-beard did you get anywhere with replacing hyper with something that is a bit more wasm friendly? I am interested in whether I can get this library working as part of a Cloudfare worker...

**Comment by arn-the-long-beard** [comment]
Hey @allsey87  :smile: 

Thank you for your message.

Well, my programmings skills are not gonna get improved because I am moving to more project management and less coding. I created a Youtube Api in Rust-Wasm because I do not need the other google Apis for now. This is how I end up with a solution for now, staying simple and scope & focus only on the part of the Api I needed.

Unfortunately I did not spend more time on this repository specifically because it would have been to difficult for me to make any change for wasm compatibility.

The Youtube api I have is only for the framework https://github.com/seed-rs/seed .

**Reply by allsey87** [reply]
Yeah, I think this is the avenue I might end up taking -- just extract what I need, swapping out the dependencies on hyper and tokio etc. for the components from [workers-rs](https://github.com/cloudflare/workers-rs) that implement similar functionality.

It's a shame that there isn't some standard traits for making web requests that are independent of tokio and hyper.
